lib/mailchimp.rb

```
@@ -0,0 +1,69 @@
+require 'net/http'
+require 'json'
+
+module BeyondZ
+  class Mailchimp
+    attr_reader :email
+    
+    HOST = "https://us11.api.mailchimp.com"
+    VERSION = '3.0'
+    
+    def initialize email
+      @key = ENV['MAILCHIMP_KEY']
+      @list_id = ENV['MAILCHIMP_LIST']
+      puts "key: #{@key}, list: #{@list_id}"
+      @email = email
+    end
+    
+    def update new_email
+      return false unless exists?
+      
+      subscriber_id = search['exact_matches']['members'].first['id']
+      
+      uri = URI("#{HOST}/#{VERSION}/lists/#{@list_id}/members/#{subscriber_id}")
+      
+      request = Net::HTTP::Patch.new(uri)
+      request.basic_auth 'key', @key
+      request.body = {email_address: new_email}.to_json
+      
+      response = http.request request
+      
+      json = JSON.parse(response.body)
+      json['email_address'] == new_email
+    rescue
+      false
+    end
+    
+    private
+    
+    def exists?
+      search['exact_matches']['total_items'].to_i > 0
+    rescue
+      false
```

Should we log or do something that makes it easier to troubleshoot when the HTTP request fails?

---

I based this off our pre-existing custom LMS class, and added some graceful fault tolerance from there. Worst case, it prevents the user record from being updated. As we talked about in the standup, this does mean that staging user e-mail updates will usually fail, unless we have a corresponding subscriber in the staging list on mailchimp. Oh, and we need an official staging list on mailchimp.

I think that if the mailchimp update fails, maybe we should go ahead and refuse to update the record, like I coded it. BUT, I think we should throw an error on the ActiveRecord object that will float up through the UI and tell the user *why* the update didn't go through. This provides clarity while maintaining data integrity. @adamdruppe @sadleb Thoughts?

---

That sounds like a good plan. This is a staff facing tool, so having them see exceptions is good.